[Feat]: Notification List UI Implementation

Ayush8923 opened the pull request:
closes: https://github.com/Aam-Digital/ndb-core/issues/2661

### Visible/Frontend Changes
-  Implement the Notification List UI as per this [comment](https://github.com/Aam-Digital/ndb-core/issues/2661#issuecomment-2531246221).
- Create the different components for the `Notification-items`.
- Create Two tabs for all notifications and unread notifications.
- Update the notification settings icon to display three dots instead of the settings icon, and show the settings menu items upon clicking the dots.

Ayush8923 commented:
For reference, here is the notification list UI.

1. When the All notification tab is active, then we can show all the notifications.
<img width="311" alt="image" src="https://github.com/user-attachments/assets/c50c10aa-9717-465f-b078-9f488c966ad0">

2. If there is no notification for the user.
<img width="295" alt="image" src="https://github.com/user-attachments/assets/f72ed0af-f6cf-4afa-937a-3f14a217cbc6">

3. When the user clicks on the three dots, the menu will offer options to mark the notification as read or delete it.
<img width="252" alt="image" src="https://github.com/user-attachments/assets/9d9ee7f7-9c38-46c3-8334-505c1c7c2293">

4. When the user clicks on the header settings menu, they can mark the notification as active and also be redirected to the notification settings page.
<img width="250" alt="image" src="https://github.com/user-attachments/assets/f73f9351-7106-465c-9d3d-05228313930e">

Ayush8923 commented:
In this PR, the notification list is dynamic. I have removed the hardcoded notification list.

Ayush8923 commented on src/app/features/notifications/notifications.component.ts:
```diff
@@ -94,18 +85,51 @@ export class NotificationsComponent implements OnInit {
+    ).length;
+  }
+
+  markAllRead($event: Event) {
+    $event.stopPropagation();
+    Logging.log("All notifications marked as read");
+  }
```
Need to update the logic so that when the user click on the mark all as read button then all the notification mark as read.

sleidig replied:
Can you implement that now already to finalize this component? (I guess updating all unread notifications and then saving those using the entityMapper?)

Ayush8923 replied:
For these types of functionalities, I have created this [issue](https://github.com/Aam-Digital/ndb-core/issues/2690). Can we implement this functionality under that issue?

Ayush8923 commented on src/app/features/notifications/notifications.component.ts:
```diff
@@ -94,18 +85,51 @@ export class NotificationsComponent implements OnInit {
+    Logging.log("All notifications marked as read");
+  }
+
+  markAsRead(notification: NotificationActivity): void {
+    // Need to add/update this logic to mark as read the notification from the CouchDB
+    notification.readStatus = true;
+    this.hasUnreadNotificationCount = this.countUnreadNotifications(
+      this.allNotifications,
+    );
+    Logging.log("Notification marked as read");
+  }
```
This function is for when the user mark as read in single notification.

sleidig replied:
I think you can simple replace the Logging with a call to EntityMapperService saving the updated notification entity?
`await this.entityMapper.save(notification);`

Ayush8923 commented on src/app/features/notifications/notifications.component.ts:
```diff
@@ -94,18 +85,51 @@ export class NotificationsComponent implements OnInit {
+    Logging.log("Notification marked as read");
+  }
+
+  deleteNotification(notification: NotificationActivity) {
+    // Need to add/update this logic to delete the notification from the CouchDB
+    this.allNotifications = this.allNotifications.filter(
+      (n) => n !== notification,
+    );
+    this.unreadNotifications = this.unreadNotifications.filter(
+      (n) => n !== notification,
+    );
+    this.hasUnreadNotificationCount = this.countUnreadNotifications(
+      this.allNotifications,
+    );
+    Logging.log("Notification deleted");
```
This function is used for if user need to delete any notification.

Ayush8923 commented:
@sleidig This PR is ready for review. Can you please review this PR and let me know if there any changes needed.

Ayush8923 commented on src/app/features/notifications/notifications.component.ts:
```diff
@@ -74,16 +66,15 @@ export class NotificationsComponent implements OnInit {
-    this.notifications = this.filterUserNotifications(
-      allNotifications,
-      "User:demo",
+    this.filterUserNotifications(notifications, "User:demo");
+    this.hasUnreadNotificationCount = this.countUnreadNotifications(
+      this.allNotifications,
     );
```
This is to get the dynamic notification list corresponding to the user. But the `this.currentUser.value?.getId()` getting undefined that's why I have added the hardcoded value just for testing purposes. I will remove this when I implement the notification list logic.

sleidig replied:
I think in our case we should work with `SessionInfo` and its `id` property to identify and filter for users. Every authenticated user will have this (it's the keycloak user account ID).

While the demo should give you a defined user entity with `CurrentUserSubject` (not sure why you are getting undefined ...), this is not available for every user in every system, depending on the custom setup.

sleidig commented on src/app/features/notifications/model/notifications-activity.ts:
```diff
@@ -0,0 +1,18 @@
+import { DatabaseField } from "../../../core/entity/database-field.decorator";
+import { DatabaseEntity } from "../../../core/entity/database-entity.decorator";
+
+@DatabaseEntity("Notification")
+export class NotificationActivity extends Entity {
```
```suggestion
/**
 * This represents one specific notification event for one specific user,
 * displayed in the UI through the notification indicator in the toolbar.
 */
@DatabaseEntity("NotificationEvent")
export class NotificationEvent extends Entity {
```

Let's discuss the most clear naming. I am thinking:
- NotificationEvent = single, specific notification entry
- NotificationRule (or maybe NotificationSubscription or ....) = one configuration of a certain notification type for which the user wants to see events
- ... maybe there are other entities or interfaces also in the overall picture but I can't think of any right now.

What do you think? Please suggest alternative namings also.

sleidig commented on src/app/features/notifications/model/notifications-activity.ts:
```diff
@@ -0,0 +1,18 @@
+
+@DatabaseEntity("Notification")
+export class NotificationActivity extends Entity {
+  @DatabaseField({ dataType: "string" }) title: string;
+  @DatabaseField({ dataType: "string" }) body: string;
+  @DatabaseField({ dataType: "string" }) actionURL: string;
+
+  @DatabaseField({ dataType: "string" }) sentBy: string;
+  @DatabaseField({ dataType: "string" }) fcmToken: string;
+  @DatabaseField({ dataType: "boolean" }) readStatus: boolean;
```
You can skip the dataType (just put `@DatabaseField()`) unless it is something special. The annotation infers it automatically from the Typing of the property

sleidig reviewed (commented):
Overall approach looks good :+1:

Ayush8923 commented on src/app/app.component.ts:
```diff
@@ -25,11 +25,7 @@ import { filter } from "rxjs/operators";
  */
 @Component({
   selector: "app-root",
-  template: `@if (configFullscreen) {
-      <router-outlet></router-outlet>
-    } @else {
-      <app-ui></app-ui>
-    }`,
```
For this, I have created the `app.component.html` file and used it directly here.

sleidig replied:
Is this directly related to your PR?

If not, let's not do it as part of this branch :blush:
Separate template file may be preferable and this is a simple change but having unrelated code has sometimes bloated PRs or even led to side-effects previously, so let's be strict with "one task, one PR"

Ayush8923 commented:
@sleidig The PR deployment is failing, possibly due to insufficient Docker space. Could you please take a look into this? Once resolved, Neha will test the UI and share any feedback/suggestions.

Ayush8923 commented:
@george-neha I have added the mocked data for UI testing purposes, please test the notification list UI and share your feedback/suggestions.

Ayush8923 commented:
> Deployed to https://pr-2711.aam-digital.net/

@george-neha I have added the mocked data for UI testing purposes, please test the notification list UI in the above-mentioned URL and share your feedback/suggestions.

george-neha commented:
The UI for this notification feature is working as per functionality. However, the timestamp for each notification is a little confusing. 
<img width="960" alt="image" src="https://github.com/user-attachments/assets/d9fbf7bc-5107-4588-b5c9-effd52466360" />
**Suggestion:** The notifications can be segregated into "Now/New" and "Earlier." 
<img width="292" alt="image" src="https://github.com/user-attachments/assets/8e75022b-de7d-4585-81b9-e704e47df714" />

The timestamp can also be displayed in this manner: 
<img width="295" alt="image" src="https://github.com/user-attachments/assets/ac8e1f5e-e0f2-49de-a8e2-043026f98313" />

sleidig commented:
> The UI for this notification feature is working as per functionality. However, the timestamp for each notification is a little confusing.

Agreed :+1:   Let's display the time since event as suggested + a tooltip on that to see the actual date and time (in human-readable, locale formatting)

sleidig commented on src/app/features/notifications/notifications.component.html:
```diff
@@ -0,0 +1,91 @@
+<button
+  mat-icon-button
+  matTooltip="Notifications"
+  i18n-matTooltip="Notifications"
```
```suggestion
  i18n-matTooltip="notifications toolbar icon tooltip"
```

sleidig commented on src/app/features/notifications/notifications.component.ts:
```diff
@@ -0,0 +1,134 @@
+    this.loadAndProcessNotifications();
+  }
+
+  onNotificationBellClick() {
+    // this.firebaseNotificationService.sendNotification();
+    Logging.log("notificationBellClicked");
+  }
```
does this still / in the future do anything? Or will the click simply open the menu (done directly in the template)?

Ayush8923 replied:
Not needed. I have used this for testing purpose. But now I have removed this.

sleidig commented on src/app/features/notifications/notifications.component.html:
```diff
@@ -0,0 +1,91 @@
+  </span>
+</button>
+
+<mat-menu #data>
```
maybe clearer?
```suggestion
<mat-menu #notificationList>
```

sleidig commented on src/app/features/notifications/notifications.component.html:
```diff
@@ -0,0 +1,91 @@
+    <div
+      class="flex-row justify-space-between align-center notification-panel-header"
+    >
+      <h1 class="notification-title" i18n>Notifications</h1>
```
```suggestion
      <h2 class="notification-title" i18n>Notifications</h2>
```

sleidig commented on src/app/features/notifications/notifications.component.html:
```diff
@@ -0,0 +1,91 @@
+      <mat-tab label="Unread" i18n-label></mat-tab>
+    </mat-tab-group>
+
+    <ng-container *ngIf="selectedTab === 0">
+      <!-- All Notifications -->
+      <app-notification-item
+        *ngFor="let notification of allNotifications"
```
shall we switch to the new syntax of `@if` and `@for` here?

sleidig commented on src/app/features/notifications/notifications.component.html:
```diff
@@ -0,0 +1,91 @@
+      </div>
+    </ng-container>
+
+    <ng-container *ngIf="selectedTab === 1">
```
rather than copying the (exactly same?) template, we can probably do something like `let ... of (selectedTab === 1 ? unreadNotifications : allNotifications)`?

Ayush8923 replied:
Thanks for this suggestion. I have update the code accordingly. Please take a look once.

sleidig commented on src/app/features/notifications/notifications.component.html:
```diff
@@ -0,0 +1,91 @@
+        class="flex-column no-notification-message"
+        *ngIf="allNotifications.length === 0"
+      >
+        <fa-icon class="no-notification-icon" icon="bell"></fa-icon>
+        <span class="no-notification-text" i18n>You have no notifications</span>
```
The icon feels very massive. Lets make this a bit smaller (maybe half the size but with margin around) and remove the bold
![image](https://github.com/user-attachments/assets/aec733b0-16ba-4541-8720-13bcae1720ce)

sleidig replied:
Thanks :-)

sleidig commented on src/app/features/notifications/notifications.component.html:
```diff
@@ -0,0 +1,91 @@
+<mat-menu #settingsMenu>
+  <button mat-menu-item (click)="markAllRead($event)">
+    <fa-icon icon="bell"></fa-icon>
+    <span class="menu-option" i18n>Mark as notification active</span>
```
```suggestion
    <span class="menu-option" i18n>Mark all notifications as read</span>
```

Ayush8923 replied:
The "Mark as notification active" option is used when the user needs to enable notifications for their account. However, to mark all notifications as read in one action, I have added a one more menu item labeled "Mark all as read." When the user clicks on this option, the readStatus of all notifications will be set to true.

sleidig replied:
Ah, I looked at the code to try and understand what you mean by "Mark as notification active". 
So this is something like "Enable notifications"?

sleidig commented on src/app/features/notifications/notifications.component.ts:
```diff
@@ -0,0 +1,134 @@
+export class NotificationsComponent implements OnInit {
+  public allNotifications: NotificationEvent[] = [];
+  public unreadNotifications: NotificationEvent[] = [];
+  public hasUnreadNotificationCount = 0;
```
simply use `unreadNotifications?.length` instead?

sleidig commented on src/app/features/notifications/notifications.component.ts:
```diff
@@ -0,0 +1,134 @@
+  public allNotifications: NotificationEvent[] = [];
+  public unreadNotifications: NotificationEvent[] = [];
+  public hasUnreadNotificationCount = 0;
+  public isEnableNotification = false;
```
is this used?

Ayush8923 replied:
Yes, this used when the user enable the notification for their account. But for now, I have removed this variable. I will implement this later when I implement the logic for the notification list.

sleidig commented on src/app/features/notifications/notifications.component.ts:
```diff
@@ -0,0 +1,134 @@
+  constructor(
+    private entityMapper: EntityMapperService,
+    private mockNotificationsService: MockNotificationsService,
+    private currentUser: CurrentUserSubject,
```
--> use SessionInfo as discussed?

Ayush8923 replied:
Yes. Now, I have used the `SessionInfo`.

sleidig reviewed (changes requested):
Regardless of my many remarks, the UI and code looks pretty good!

The UI updates (update indicator upon marking read, etc.) doesn't fully work yet. Do you still want to implement this here or only later?

Ayush8923 commented on src/app/features/notifications/notification-item/notification-item.component.html:
```diff
@@ -0,0 +1,36 @@
+    i18n-matTooltip="Unread notification"
+    matBadgeColor="accent"
+    [matMenuTriggerFor]="notificationMenu"
+    (click)="$event.stopPropagation()"
```
When I remove ` $event.stopPropagation()`, clicking on any menu item causes the menu to close automatically. For example, when I click the blue dot, the new menu item list opens, but the previous one closes. In some cases, however, I don't want the menu item to close automatically. That's why I added `$event.stopPropagation()` to prevent the menu item from closing on its own.

sleidig replied:
Ah, I see ... Seems a bit tedious to put it on almost every event handler but if we need it, that's okay. (might be any way to stop this more centrally without all the repeated  stopPropagation calls?)

Ayush8923 commented on src/app/features/notifications/notification-item/notification-item.component.html:
```diff
@@ -0,0 +1,36 @@
+
+  <!-- Blue indicator for unread notifications -->
+  <span
+    *ngIf="!notification.readStatus"
```
I have added an empty circle for read notifications. When the user clicks on the icon, they will be able to delete the notification or mark it as unread.

Ayush8923 commented on src/app/features/notifications/notification-item/notification-item.component.html:
```diff
@@ -0,0 +1,36 @@
+      <fa-icon icon="trash"></fa-icon>
+      <span class="menu-option" i18n>Delete Notification</span>
+    </button>
+  </mat-menu>
```
I have added the condition-based `mark as read` and `mark as unread` options.

Ayush8923 commented on src/app/features/notifications/NotificationTimePipe.ts:
```diff
@@ -0,0 +1,43 @@
+/**
```
@sleidig Need your suggestion on this. I have created this file for the notification time. Is this the correct way to implement it, or are there any changes you would recommend?"

Ayush8923 commented:
> > The UI for this notification feature is working as per functionality. However, the timestamp for each notification is a little confusing.
> 
> Agreed 👍 Let's display the time since event as suggested + a tooltip on that to see the actual date and time (in human-readable, locale formatting)

Thanks for this suggestion @george-neha @sleidig.
I have updated the UI accordingly, also I have added the tooltip so that when the user hovers over the date, a tooltip will appear displaying the date and time.

Ayush8923 commented on src/app/features/notification/NotificationTimePipe.ts:
```diff
@@ -0,0 +1,43 @@
+/**
```
@sleidig, Please share your suggestion on this, I have created this file for the notification time. Is this the correct way to implement it, or are there any changes you would recommend?

sleidig commented on src/app/features/notification/notification.component.ts:
```diff
@@ -0,0 +1,113 @@
+    const notifications =
+      this.mockNotificationsService.getNotifications() as unknown as NotificationEvent[];
+
+    // The user is hardcoded for testing purposes, need to remove this.
+    this.filterUserNotifications(
+      notifications,
+      this.sessionInfo.value?.entityId,
```
needs to be id, not entityId (which again may not be defined for every user account)
```suggestion
    this.filterUserNotifications(
      notifications,
      this.sessionInfo.value?.id,
```

sleidig commented on src/app/features/notification/notification.component.html:
```diff
@@ -0,0 +1,89 @@
+        ></app-notification-item>
+      }
+
+      @if (
+        (selectedTab === 0 ? allNotifications : unreadNotifications).length ===
+        0
+      ) {
+        <div class="flex-column no-notification-message">
+          <fa-icon
+            class="no-notification-icon"
+            [icon]="['far', 'bell']"
+          ></fa-icon>
+          <span class="no-notification-text" i18n
+            >You have no notifications</span
+          >
+        </div>
+      }
```
we can probably use `@empty` here to save some code: https://angular.dev/api/core/@for

sleidig reviewed (commented):
Looking pretty good already.

A few more optimizations for UI & UX:
- [x] menu icons should have some width to cleanly align: 
![image](https://github.com/user-attachments/assets/8607b1d8-89bd-4770-b389-4e60c06ec910)
- [x] that menu (mark unread / delete upon blue dot) should auto-close after clicking one option, in my opinion
- [x] same for the top "..." menu. "Mark all read" should close the action menu (but keep the notifications list open).

Ayush8923 commented on src/app/features/notification/notification.component.html:
```diff
@@ -0,0 +1,89 @@
+      <mat-tab label="Unread" i18n-label></mat-tab>
+    </mat-tab-group>
+
+    @if (selectedTab === 0 || selectedTab === 1) {
```
I checked the code again, and now we don't need this condition. The `selectedTab` will either be 0 or 1. I have updated it.

Ayush8923 commented on src/app/features/notification/NotificationTimePipe.ts:
```diff
@@ -0,0 +1,43 @@
+    } else if (days >= 7) {
+      const dayOfWeek = notificationTime.toLocaleString("en-US", {
+        weekday: "short",
+      });
```
I have modified the code to implement this:
1. For notifications 7 or more days ago, show the number of days (e.g. "10d", "15d").
2. For even longer timespans (30+ days), we can display the month and year (e.g. "Jan 2024").

Ayush8923 commented:
> that menu (mark unread / delete upon blue dot) should auto-close after clicking one option, in my opinion

@sleidig I tried to implement the functionality in this way, but the `mat-menu` is not closing due to the use of `stopPropagation`. When I use `stopPropagation` in a `mat-menu` item, the `mat-menu` doesn't close. However, when I remove `stopPropagation`, both `mat-menu` close. So, I implemented it in such a way that whenever the user selects any `mat-menu` item, the `mat-menu` does not close. If the user wants to close it, they can do so. Could you please suggest a solution if you have one?

> same for the top "..." menu. "Mark all read" should close the action menu (but keep the notifications list open).

Same for this.

Ayush8923 commented:
> Deployed to https://pr-2711.aam-digital.net/

@george-neha I have made some updates to the notification list UI. Please take a look and let us know if you have any suggestions. Also, I have modified the notification time and when the user hovers in the notification time then in the tooltip the actual date and time will be shown.
1. For notifications 7 or more days ago, show the number of days (e.g. "10d", "15d").
2. For even longer timespans (30+ days), we can display the month and year (e.g. "Jan 2024").

george-neha commented:
Working as per discussed functionality and UI. Can be moved to tech review.
@Ayush8923- Can this be aligned?
<img width="960" alt="image" src="https://github.com/user-attachments/assets/a2d4589a-6f1f-4857-953a-6e483260aa3d" />

sleidig commented on src/app/features/notification/notification.component.html:
```diff
@@ -0,0 +1,97 @@
+      <mat-tab label="Unread" i18n-label></mat-tab>
+    </mat-tab-group>
+
+    <div class="flex-column gap-small">
```
Can you design the divs in a way that we only have a scrollbar within the items and the title + tabs are always visible outside of the scroll box?

![image](https://github.com/user-attachments/assets/11f6dfb2-0edb-49d1-a67c-7a9c166ebeeb)

Ayush8923 replied:
Yes, we can do this. I have updated the code accordingly. but slight text is shown in the background at the top when we scroll the notification list.

Ayush8923 replied:
It's working properly now. When the user scrolls through the notification list, only the notification list will scroll, while the title and tabs remain fixed.

sleidig replied:
somehow, if I scroll all the way to the bottom the top bar eventually scrolls out, although initially it remains fixed as intended, @Ayush8923

Ayush8923 replied:
@sleidig  I have fixed this. Now, when the user scrolls the notification list then the title and tabs remain fixed.

sleidig replied:
your fixed caused a horizontal scrollbar for me and displayed the scroll across the whole menu rather than just the part that is actually scrollable. I made some further adjustments to fix this.

sleidig reviewed (approved):
Made some small improvements. Looks fine now :+1:
